formats/json-tests/commonTest/src/kotlinx/serialization/JsonElementPolymorphicErrorTest.kt

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+
+import kotlinx.serialization.json.*
+import kotlinx.serialization.modules.SerializersModule
+import kotlinx.serialization.modules.polymorphic
+import kotlinx.serialization.test.*
+import kotlin.test.*
+
+class JsonElementPolymorphicErrorTest : JsonTestBase() {
+
+    @Serializable
+    abstract class Abstract
+
+    @Serializable
+    data class IntChild(val value: Int) : Abstract()
+
+    @Serializable
+    data class CollectionChild(val value: Int) : Abstract()
+
+    @Serializable
+    data class Holder(val value: Abstract)
+
+    private val format = Json {
+        prettyPrint = false
+        serializersModule = SerializersModule {
+            polymorphic(Abstract::class) {
+                subclass(IntChild::class, IntChildSerializer)
+                subclass(CollectionChild::class, CollectionChildSerializer)
+            }
+        }
+    }
+
+    object IntChildSerializer : JsonTransformingSerializer<IntChild>(serializer()) {
+        override fun transformSerialize(element: JsonElement): JsonElement {
```

Comment:
if I understood the question correctly, then if the input is not an object, but, for example, an array, then we will not even begin deserialization, because the check above will work.

E.g. if JSON instead of `{"value":{"type":"kotlinx.serialization.JsonElementPolymorphicErrorTest.IntChild","value":42}}` we specify `{"value": [1, 2]}`, we got error:
```
kotlinx.serialization.json.internal.JsonDecodingException: Expected JsonObject, but had JsonArray as the serialized body of kotlinx.serialization.Polymorphic<Abstract> at element: $.value
JSON input: [1,2]
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:24)
	at kotlinx.serialization.json.internal.JsonExceptionsKt.JsonDecodingException(JsonExceptions.kt:32)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableValue(StreamingJsonDecoder.kt:399)
	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableValue(AbstractDecoder.kt:43)
	at kotlinx.serialization.encoding.AbstractDecoder.decodeSerializableElement(AbstractDecoder.kt:70)
	at kotlinx.serialization.json.internal.StreamingJsonDecoder.decodeSerializableElement(StreamingJsonDecoder.kt:168)

```

Reply:
Yeah, I'm asking to write a test for this error message so we are sure that it is reasonable and understable for user. Although there is https://github.com/Kotlin/kotlinx.serialization/blob/e35c28dd0ecbf99460635059a39669683bb24cbf/formats/json-tests/commonTest/src/kotlinx/serialization/json/JsonErrorMessagesTest.kt#L25 currently, maybe it is enough